### ISSUE TITLE
Grammar versions and dealing with breaking changes

### ISSUE THREAD

**Issue by dannyfreeman** [tree-sitter/tree-sitter#2611]
Just about any time a change is made to a grammar, the change is somehow breaking to consumers of the grammar. Right now in Emacs we don't have a great way to tell what version of a grammar is loaded without resorting to writing a simple query and seeing if it throws an error.

For example, in the javascript grammar a change was made that renamed a node. In Emacs we have a special check when writing indentation rules for this https://github.com/emacs-mirror/emacs/blob/0cf8a16928841ad120c697455f28106a5cca4a09/lisp/progmodes/js.el#L3434-L3440

As grammars evolve more and more of these changes accumulate. Since tools may use grammars that are packaged by a linux distro, there is a need to support older grammar versions.

It would be nice if there was somehow a way to detect what "version" of a grammar is loaded without executing trial-by-error queries to narrow down what version of the grammar we might be working with.

Emacs has the ability to help the user download grammars using a specific git tag, but nothing compels grammar authors to even use git tags with version numbers in them. Some kind of version number baked into the generated C code I believe would be a nice solution. Even if it's just strictly increasing number every time the grammar is generated would go a long way. This would allow us to call into the dynamic library and see if it is one of the grammar versions we tested with and know works.

I am very interested in other solutions to this. How do other tools deal with these breaking changes? Is it a matter of always downloading and compiling the exact grammar version needed?

**Comment by amaanq** [comment]
The best solution for this imo is to use the tagged version, I try to keep upstream and my own grammars and those I have access to on a stable release cycle based on the tags, but not every grammar can do this.

Neovim/nvim-treesitter currently just updates ASAP (courtesy of whoever submits a PR, typically the grammar maintainer listed in the readme or me)

I'm not keen on baking a version number into the generated code..too much mismatching can occur and that's not a burden I want downstream to have, releasing git tags + publishing packages is the way to go imo

Being realistic, grammars are going to have changes from time to time, languages evolve, some code was refactored better to improve parser size/state count, naming convention might change to match say, the upstream language version, etc, there's just really no solution to solve this for downstream

I would follow how nvim-treesitter does it for emacs, to have a lockfile of grammar revisions of sorts and use a CI action to auto-update them and see if the queries break on the new version. Then you don't have to track master and wonder why a grammar broke in the future

**Comment by clason** [comment]
I would go further (and have done so repeatedly, here and elsewhere): for tree-sitter to seriously support this use case (having parsers and queries hot-swappable, user-customizable "runtime files"), there _needs_ to be some introspection that allows tying queries to a specific grammar version. The lack of this is (together with an official binary distribution strategy) one of the major pain points in the ecosystem.

Until then, this is in the hands of the ecosystem maintainers (Neovim, Emacs, Helix) to enforce. The current Neovim strategy is pretty much YOLO (we have an automated action that bumps the parser commits we track in our distribution plugin, nvim-treesitter, and merge it if basic tests pass), but that is clearly not sustainable due the volume of changes. 

So I've been trying to push grammar maintainers to actually follow a **semver-style release**, where 
1. point updates are grammar fixes that don't affect queries at all;
2. minor updates add grammar features (nodes) that need query updates to make use of, but queries work fine with otherwise;
3. major updates remove/change grammar features (nodes) that break current queries.

Once that becomes widespread, we will change the strategy to prefer versions over commits -- which then announce compatibility themselves so this information can be passed on to users ("dependabot-style").

**Comment by dgutov** [comment]
> I'm not keen on baking a version number into the generated code..too much mismatching can occur and that's not a burden I want downstream to have, releasing git tags + publishing packages is the way to go imo

This would be the obvious choice if not for the Linux distros already packaging and distributing ts grammars.

When those are available, it kind of makes sense to try to use them (e.g. to avoid asking *every* user to build the grammar during installation, with associated toolchain requirements), and then we're back to having to support a range of versions.

**Comment by pokey** [comment]
I believe the discussion in https://github.com/tree-sitter/tree-sitter/discussions/1768 may be relevant fwiw

**Comment by clason** [comment]
>  "this grammar was generated on this day, so it should support querying nodes X, Y, and Z" with a good degree of confidence.

No, sorry, that's not good enough, since grammars can -- and do -- make multiple breaking changes per day. LuaJIT works around this with a "seconds since epoch" timestamp, but that is not an example I would like to see followed, as it makes it very hard to reason about. 

Furthermore, not every silly "typofix" or "regenerate files" commit needs to lead to a breaking change, and versioning is exactly what is used everywhere else to convey this semantic information. Why reinvent the wheel?

(I don't quite see the issue with Emacs, to be honest -- Neovim's "luxury" is the problem here, and if you bundle both queries and parsers, it should be straightforward to update them together -- or not, if you prefer to be on the safe side? But this is something that is better discussed offline; feel free to reach out on Matrix if you'd like to compare notes with other editors on parser and query management (which we would be very happy about; better downstream coordination something I feel strongly about).)

**Comment by clason** [comment]
> That is the thing, Emacs only bundles queries, not parsers. Users install the parsers. OS distros are also packaging parsers and Emacs will try to use them if they are present

Ah, I see. Yes, there's a definite push for that, but it's just not workable given the state of things; we push back hard on that, and our strategy -- which I strongly recommend -- is reversing the priority: user config > plugins > bundled > system (which are never looked at by default in Neovim). 

(Just to be clear: I agree that some sort of introspection is urgently needed, for the very reason you describe. I just disagree -- strongly -- that a timestamp is better than explicit, semantic, versioning.)

**Comment by pokey** [comment]
Fwiw in Cursorless we pin to specific commit SHAs of grammar repos, and then build wasms and bundle them into our extension. Obviously not ideal, but does allow us to be sure all our queries will work, and allows us to track grammars that release infrequently by using SHAs rather than published grammars

**Comment by clason** [comment]
That's precisely how nvim-treesitter works -- with a daily workflow that tries to update to the latest commit of each of the ~230 tracked parsers and runs tests; if they all come up green, we merge. This has the obvious drawback that we can miss _additions_ to the grammar that could be used to improve queries. We rely on "language maintainers" to track them and make followup PRs to the queries; obviously that scales poorly... This is another issue that I hope semver will alleviate.

**Comment by jeff-hykin** [comment]
> Is it a matter of always downloading and compiling the exact grammar version needed?

Yep. Using specific wasm files to be exact.

**Comment by jeff-hykin** [comment]
I think there could be a really good objective/automatic semver to address the larger issue.

I'd support an MVP (some kind of number in generated grammars)

But I think a great system would be the following:
- *major version increment:* deletion of a tag name, attribute name, or a removal of a hierarchy option (e.g. `(statement (wrapper (thing` becoming `(statement (thing`) 
- *minor version increment:* addtion of a tag,  hierarchy, or attribute 
- *patch:* fixing a case that would have otherwise been an error (this would be undetectable by an automated scanner, but thats fine so long as the other two cases are detectable)

I think such a system would work well for syntax highlighters and such. ≥minor version increments would be compatible, and any patch number would be acceptable.

**Comment by clason** [comment]
So basically exactly what I wrote in https://github.com/tree-sitter/tree-sitter/issues/2611#issuecomment-1709647549.

**Comment by jeff-hykin** [comment]
Mostly, I'm saying I think the pack/publish command of the tree sitter CLI could evemtaully automate this instead of asking grammar maintainers to do their best.

**Comment by ObserverOfTime** [comment]
PR welcome. Good luck.
